app/queries/consumer_apps/find_or_create_all_query.rb

```diff
@@ -5,7 +5,8 @@ def self.call
       existing_platforms = ConsumerApp.where(app_bundle: forem_bundle).pluck(:platform)
       (ConsumerApp::FOREM_APP_PLATFORMS - existing_platforms).each do |platform|
         # Re-create the supported Forem apps if they're missing
-        ConsumerApp.create(app_bundle: forem_bundle, platform: platform, active: true)
+        ConsumerApp.create(app_bundle: forem_bundle, platform: platform, active: true,
```

Comment:
Interesting, I'm not sure I understand what type of failure handling we would want to have here.

1. We are trying to "create" the `ConsumerApp`s that are missing (i.e. Forem's iOS `ConsumerApp`)
2. If the create fails because a validation on the `ConsumerApp` model we should technically catch that in CI before introducing the failure in the first place with: [find_or_create_all_query_spec](https://github.com/forem/forem/blob/main/spec/queries/consumer_apps/find_or_create_all_query_spec.rb) & [find_or_create_by_query_spec](https://github.com/forem/forem/blob/main/spec/queries/consumer_apps/find_or_create_by_query_spec.rb)

This is a query, so in case the "create" fails the result will continue to be an AR relation (the result of `ConsumerApp.limit(50)`) but with fewer than expected results (i.e. missing the failed ConsumerApp that failed to be created here). We can add a bang to it `ConsumerApp.create!(app_bundle: ...` and then it would fail on us whenever performing the query, clearly bubbling up the hypothetical problem by "crashing" the Consumer App UI.

I can add the bang for the query to raise an exception if the create fails, but I just want to make sure we're both on the same page of what this failure handling would mean 🙂 

An alternative I can think of is for example: Add the bang (`ConsumerApp.create!(app_bundle: ...`) but wrap it in a `begin-rescue` clause and then log to Datadog/Honeybadger the error but continue to "fail silently" on the UI

Reply:
@fdoxyz 

> I'm not sure I understand what type of failure handling we would want to have here.

Yeah, that was pretty vague because I had no idea what we'd want there, either. :joy:

> in case the "create" fails the result will continue to be an AR relation (the result of `ConsumerApp.limit(50)`) but with fewer than expected results (i.e. missing the failed ConsumerApp that failed to be created here). We can add a bang to it `ConsumerApp.create!(app_bundle: ...` and then it would fail on us whenever performing the query, clearly bubbling up the hypothetical problem by "crashing" the Consumer App UI.

Yeah, sounds like this is a bit more complicated than I hoped and the potential of silent failure already exists, so maybe this discussion belongs in its own PR. I'm very much not the judge of what makes the most sense in this case, but when I see an ActiveRecord `save`, `create`, or `update` without checking the return value this is typically where my mind goes.

A few ways I've typically handled stuff like this, with tradeoffs between them:

- Graceful degradation with ActiveModel validation errors in the UI
  - 👍🏼 Rails makes this easy for a given record
  - 👎🏼 I don't know of an easy way to handle this for many records, which it looks like is what we're dealing with here
  - 👎🏼 Really only works well if the model is just a persisted version of the HTML form
- Don't save inside the web request. Kick it off to a Sidekiq job instead.
  - 👍🏼 You can raise the exception and the end user's workflow is not broken. From their perspective it just takes longer. This is extremely common in CQRS application architectures.
  - 👎🏼 If the code is fine but the data you're trying to store is wrong, you may not be able to fix that
- When in doubt, scream it out — bang method, raise the exception.
  - 👍🏼 We get a record of what happened, who it happened to, where it happened, and all the stack frames underneath it
  - 👎🏼 It's an absolute shit UX

I know I said this discussion probably belongs in another PR, but it was in my head now and I probably wouldn't remember it well later. :joy:

Reply:
Awesome, thank you for the input @jgaskins! I appreciate the thoughts because it's something we can always look back to and reference regardless if we consider it for this PR or for a bigger discussion 😃 

For the record: I'm leaning a bit more towards the last option because this is a very unlikely event and under our control (not user input related for example) that should ideally be caught in CI (the situation where these records need to be created). Also, the other two options would require a bit more of a refactor in how these queries work and in that case I'd definitely push for doing that work in a separate PR.

No strong opinions from my end though, I'm okay leaving it as-is or adding the bang to raise an exception if someone else also feels the PR would be better off if we include it.

Reply:
I'm good with option 1 or 3 as well! Maybe we can use `create!` and then log the error to Datadog, so that we can monitor to see if this actually happens in production and we'll know why.

How does that sound to you @fdoxyz and @jgaskins ?

Reply:
Definitely sounds good to me! I'll include this to the PR 👍